tiledb/sm/fragment/test/unit_fragment_identifier.cc

```diff
@@ -0,0 +1,189 @@
+      {"__0123456789ABCDEF0123456789ABCDEF_1_2_5"},   // __uuid_t1_t2_v
+      {"_____120123456789ABCDEF0123456789ABCDEF5"},   // _____t1t2uuidv
+      {"____1_20123456789ABCDEF0123456789ABCDEF5"},   // ____t1_t2uuidv
+      {"___1__20123456789ABCDEF0123456789ABCDEF5"},   // ___t1__t2uuidv
+      {"__1___20123456789ABCDEF0123456789ABCDEF5"},   // __t1___t2uuidv
+      {"_1____20123456789ABCDEF0123456789ABCDEF5"},   // _t1____t2uuidv
+      {"1____20123456789ABCDEF0123456789ABCDEF5"},    // t1_____t2uuidv
+      {"1___2_0123456789ABCDEF0123456789ABCDEF5"},    // t1____t2_uuidv
+      {"1__2__0123456789ABCDEF0123456789ABCDEF5"},    // t1__t2___uuidv
+      {"1_2___0123456789ABCDEF0123456789ABCDEF5"},    // t1_t2___uuidv
+      {"1_2__0123456789ABCDEF0123456789ABCDEF_5"},    // t1_t2__uuid_v
+      {"1_2_0123456789ABCDEF0123456789ABCDEF__5"},    // t1_t2_uuid__v
+      {"1_20123456789ABCDEF0123456789ABCDEF___5"},    // t1_t2uuid___v
+  }};
+
+  // Timestamps and uuid are identical
+  std::vector<failure_test_case> uuid_timestamps{{
+      {"__0123456789ABCDEF0123456789ABCDEF_"
+       "0123456789ABCDEF0123456789ABCDEF"},  // version1: __uuid_t1
+      {"__0123456789ABCDEF0123456789ABCDEF_"
+       "0123456789ABCDEF0123456789ABCDEF_"
+       "0123456789ABCDEF0123456789ABCDEF"},  // version1: __uuid_t1_t2,
+                                             // versions 2,3:
+                                             // __t1_t2_uuid
+      {"__0123456789ABCDEF0123456789ABCDEF_"
+       "0123456789ABCDEF0123456789ABCDEF_0123456789ABCDEF0123456789ABCDEF_"
+       "5"},  // version3: __t1_t2_uuid_v
+  }};
+
+  std::vector<std::vector<failure_test_case>> failure_cases{
+      invalid_uris,
+      empty_fields,
+      two_fields,
+      three_fields,
+      four_fields,
+      uuid_timestamps};
+
+  for (auto failure_case : failure_cases) {
+    for (auto failure : failure_case) {
+      auto uri = frag_dir + failure.path;
+      DYNAMIC_SECTION(uri) {
+        REQUIRE_THROWS(FragmentID{uri});
+      }
+    }
+  }
+}
+
+// Should succeed and do
+TEST_CASE("FragmentID: Valid uris", "[fragment_id][valid_uri]") {
+  std::array<success_test_case, 5> success_cases{{
+      {"file:///__0123456789ABCDEF0123456789ABCDEF_1",
+       "__0123456789ABCDEF0123456789ABCDEF_1",
+       std::pair{1, 1},
+       1,
+       2},
+      {"file:///__0123456789ABCDEF0123456789ABCDEF_1_2",
+       "__0123456789ABCDEF0123456789ABCDEF_1_2",
+       std::pair{2, 2},
+       1,
+       2},
+      {"file:///__0123456789ABCDEF0123456789ABCDEF_2_1",
+       "__0123456789ABCDEF0123456789ABCDEF_2_1",
+       std::pair{1, 1},
+       1,
+       2},
+      {"file:///__1_2_0123456789ABCDEF0123456789ABCDEF",
+       "__1_2_0123456789ABCDEF0123456789ABCDEF",
+       std::pair{1, 2},
+       2,
+       4},
+      {"file:///__1_2_0123456789ABCDEF0123456789ABCDEF_5",
+       "__1_2_0123456789ABCDEF0123456789ABCDEF_5",
+       std::pair{1, 2},
+       3,
+       5},
+  }};
```

Comment:
We don't have a test vector for a fragment-format version 3 identifier with an optional array-format identifier at the end. At least it's documented as optional.
Not making this a blocker, but it's an inconsistency between the documentation and the tests.